// teseract/lib/commands/serve.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var pathUtils = __importStar(require("path"));
var http_1 = __importDefault(require("http"));
var chokidar_1 = __importDefault(require("chokidar"));
var mime_types_1 = __importDefault(require("mime-types"));
var builder_1 = __importDefault(require("../builder"));
var child_process_1 = require("child_process");
var config_1 = require("../config");
function default_1() {
    return __awaiter(this, void 0, void 0, function () {
        var config, watcher, server, tscProcess;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, (0, config_1.getConfig)()];
                case 1:
                    config = _a.sent();
                    return [4 /*yield*/, watch(p([process.cwd(), 'src']))];
                case 2:
                    watcher = _a.sent();
                    watcher.on('all', function () {
                        resourcesPromise = null;
                    });
                    server = http_1["default"].createServer(function (req, res) { return __awaiter(_this, void 0, void 0, function () {
                        var path, _a, resources, errors, matchingResources, matchingResource;
                        return __generator(this, function (_b) {
                            switch (_b.label) {
                                case 0:
                                    if (req.url == null) {
                                        res.statusCode = 404;
                                        res.end();
                                        return [2 /*return*/];
                                    }
                                    path = req.url.endsWith('/')
                                        ? "".concat(req.url, "index.html")
                                        : req.url;
                                    return [4 /*yield*/, getResources()];
                                case 1:
                                    _a = _b.sent(), resources = _a.resources, errors = _a.errors;
                                    if (errors.length > 0) {
                                        res.statusCode = 200;
                                        res.setHeader('Content-Type', 'text/plain; charset=utf-8');
                                        res.end(errors.map(function (e) { return "".concat(e.file, "\n\n").concat(e.text, "\n\n\n"); }).join(''));
                                        return [2 /*return*/];
                                    }
                                    matchingResources = resources.filter(function (r) { return r.path === path; });
                                    if (matchingResources.length === 0) {
                                        res.statusCode = 404;
                                        res.end();
                                        return [2 /*return*/];
                                    }
                                    matchingResource = matchingResources[0];
                                    res.setHeader('Content-Type', mime_types_1["default"].lookup(path) || 'application/octet-stream');
                                    res.end(matchingResource.data);
                                    return [2 /*return*/];
                            }
                        });
                    }); });
                    server.listen(config.serve.port, config.serve.host, function () { });
                    tscProcess = startTSCProcess();
                    return [4 /*yield*/, waitTSC(tscProcess)];
                case 3:
                    _a.sent();
                    return [4 /*yield*/, closeHttpServer(server)];
                case 4:
                    _a.sent();
                    watcher.close();
                    return [2 /*return*/];
            }
        });
    });
}
exports["default"] = default_1;
var resourcesPromise = null;
function getResources() {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            if (resourcesPromise)
                return [2 /*return*/, resourcesPromise];
            return [2 /*return*/, resourcesPromise = buildResources()];
        });
    });
}
function buildResources() {
    return __awaiter(this, void 0, void 0, function () {
        var err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, (0, builder_1["default"])()];
                case 1: return [2 /*return*/, _a.sent()];
                case 2:
                    err_1 = _a.sent();
                    console.log(err_1);
                    return [2 /*return*/, { resources: [], errors: [] }];
                case 3: return [2 /*return*/];
            }
        });
    });
}
function watch(path) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (resolve) {
                    var watcher = chokidar_1["default"].watch(path);
                    watcher.on('ready', function () { return resolve(watcher); });
                })];
        });
    });
}
function startTSCProcess() {
    return (0, child_process_1.spawn)('tsc', ['--watch', '--noEmit', '--project', '.'], { stdio: 'inherit' });
}
function waitTSC(tscProcess) {
    return new Promise(function (resolve, reject) {
        tscProcess.on('close', function () {
            resolve(null);
        });
    });
}
function closeHttpServer(server) {
    return new Promise(function (resolve, reject) {
        server.close(function (err) {
            if (err) {
                reject(err);
                return;
            }
            resolve(null);
        });
    });
}
var p = function (chunks) { return pathUtils.join.apply(pathUtils, chunks); };
